### DISCUSSION TITLE
Remove ActiveMerchant dependency from Solidus

### DISCUSSION THREAD

**Discussion by jordan-brough** [solidusio/solidus#4612] [New Features or Ideas]
I don't see a reason to [require ActiveMerchant in Solidus](https://github.com/solidusio/solidus/search?utf8=%E2%9C%93&q=ActiveMerchant&type=) and it seems like it would make other billing integrations easier if we provide our own classes for errors/responses/etc.

**Comment by tvdeyen** [comment]
💯✖️👍

**Comment by jhawthorn** [comment]
One of the few things preventing this is our use of [`ActiveMerchant::Billing::Response`](https://github.com/solidusio/solidus/search?l=Ruby&q=ActiveMerchant&type=&utf8=%E2%9C%93). We can probably introduce our own class, which quacks the same way (duplicate the interface it has with duck-typing)

**Comment by Dkendal** [comment]
I'm going to take a stab at this

**Comment by Dkendal** [comment]
As far as I can see this requires: 
- Replace ActiveMerchant::BillingResponse
- Replace ActiveMerchant::ConnectionError
- replace usages for ActiveMerchant::Billing::Base??

Not sure yet what `ActiveMerchant::Billing::Base` is used for, I can only find assignments to it, never anything reading from it, including, or extending the module.

**Comment by mamhoff** [comment]
If it isn't used and you can remove it from core without anything breaking, remove it.

**Comment by Dkendal** [comment]
https://github.com/solidusio/solidus/compare/master...Dkendal:feature/remove-active-merchant?expand=1

Current branch.

**Comment by Dkendal** [comment]
@mamhoff any thoughts on how to rectify the interface with solidus_gateway? 

All results that come back from gateway are handled in `core/app/models/spree/payment/processing.rb`, I was thinking of checking the type of the result and presenting a deprecation notice if it's an `ActiveMerchant::Billing::Response`.

Thoughts?

**Comment by Dkendal** [comment]
#1935
